### PR TITLE
DRAFT: MaterialEditor support for editor_test.py framework

### PR DESCRIPTION
**Pull request by jromnoa**
- Draft PR for helping to diagnose a bug with `.get_returncode()` in the launcher code for LyTestTools.
- You will need to "Load Diff" the `editor_test.py` code because the module has a lot of lines.
- Unit tests will be added next (not complete).

### CODE REVIEW

File: AutomatedTesting/Gem/PythonTests/Atom/TestSuite_Main_GPU.py [hunk 1]
```diff
@@ -154,35 +153,19 @@ def wrap_run(self, request, workspace, editor, editor_test_results, launcher_pla
...
-    @pytest.mark.test_case_id("C30973987")  # Material Editor Launching in Vulkan
-    def test_MaterialEditorLaunch_AllRHIOptionsSucceed(
-            self, request, workspace, project, launcher_platform, generic_launcher, exe_file_name, cfg_args,
-            expected_lines):
-        """
-        Tests each valid RHI option (Null RHI excluded) can be launched with the MaterialEditor.
-        Checks for the specific expected_lines messaging for each RHI type.
-        """
+@pytest.mark.parametrize("launcher_platform", ['windows_material_editor'])
+class TestMaterialEditor(EditorTestSuite):
```
Comment (AlexOteiza) [inline AutomatedTesting/Gem/PythonTests/Atom/TestSuite_Main_GPU.py:157]: Better call this MaterialEditorTestSuite

File: Tools/LyTestTools/ly_test_tools/o3de/editor_test.py [hunk 2]
```diff
@@ -268,7 +293,7 @@ def __init__(self, test_spec: type(EditorTestBase), output: str, time_secs: floa
             self.test_spec = test_spec
             self.time_secs = time_secs
             self.editor_log = editor_log
```
Comment (AlexOteiza) [inline Tools/LyTestTools/ly_test_tools/o3de/editor_test.py:295]: this is no longer editor_log, it should be test_app_log

File: Tools/LyTestTools/ly_test_tools/o3de/editor_test.py [hunk 3]
```diff
@@ -878,19 +943,19 @@ def _exec_editor_multitest(self, request: _pytest.fixtures.FixtureRequest,
...
         output = ""
         editor_log_content = ""
         try:
-            editor.wait(self.timeout_editor_shared_test)
-            output = editor.get_output()
-            return_code = editor.get_returncode()
+            test_launcher.wait(self.timeout_editor_shared_test)
+            output = test_launcher.get_output()
+            return_code = test_launcher.get_returncode()
             editor_log_content = editor_utils.retrieve_editor_log_content(run_id, log_name, workspace)
```
Comment (AlexOteiza) [inline Tools/LyTestTools/ly_test_tools/o3de/editor_test.py:959]: all of these should be replaced with a generic function to retrieve the current test app log

File: AutomatedTesting/Gem/PythonTests/Atom/TestSuite_Main_GPU.py [hunk 4]
```diff
@@ -154,35 +153,19 @@ def wrap_run(self, request, workspace, editor, editor_test_results, launcher_pla
...
+    # Remove -BatchMode from global_extra_cmdline_args since we need rendering for these tests.
+    global_extra_cmdline_args = []
+    use_null_renderer = False
+    material_editor_test = True
```
Comment (AlexOteiza) [inline AutomatedTesting/Gem/PythonTests/Atom/TestSuite_Main_GPU.py:161]: Don't add a bool here and add special cases. 
What if we have another app to test? We would be creating bools everywhere and make this unmaintainable.

The RFC was about implementing a base TestSuite class where the EditorTestSuite extends and sets the test app to be editor.exe. Then for material editor test we would have MaterialEditorTestSuite that extends from the base test suite and sets the app to be MaterialEditor.exe.

File: Tools/LyTestTools/ly_test_tools/o3de/editor_test.py [hunk 5]
```diff
@@ -322,6 +347,8 @@ class EditorTestSuite:
     use_null_renderer = True
     # Maximum time for a single editor to stay open on a shared test
     timeout_editor_shared_test = 300
+    # Marks the test as a MaterialEditor test instead of an Editor test.
+    material_editor_test = False
```
Comment (AlexOteiza) [inline Tools/LyTestTools/ly_test_tools/o3de/editor_test.py:351]: Commented above, this is not extensible.

File: Tools/LyTestTools/ly_test_tools/o3de/editor_test.py [hunk 6]
```diff
@@ -362,6 +389,28 @@ def _editor_test_data(self, request: _pytest.fixtures.FixtureRequest) -> EditorT
...
+        """
+        Fixture that determines which launcher to use for the current test.
+        :request: The Pytest request object.
+        :editor: The editor Launcher object.
+        :material_editor: The material_editor Launcher object.
+        :returns: Either the editor or material_editor Launcher object.
+        """
+        return self._test_launcher(request, editor, material_editor)
+
+    def _test_launcher(self, request, editor, material_editor):
```
Comment (AlexOteiza) [inline Tools/LyTestTools/ly_test_tools/o3de/editor_test.py:406]: Its not the best practice to add unit testing code inside the functionality, better move it away if possible. Can we do this differently?

**Review by AlexOteiza — CHANGES_REQUESTED** [review]
Hi,

I'm not sure if this is supposed to be the final version, but this is not the way it should be done.
This is not what the RFC proposed about creating a base TestSuite class and EditorTestSuite and MaterialEditorTestSuite childs. This adds a bool for material editor which is not extensible and adds a lot of tech debt.

File: Tools/LyTestTools/ly_test_tools/o3de/editor_test.py [hunk 7]
```diff
@@ -774,6 +830,9 @@ def _exec_editor_test(self, request: _pytest.fixtures.FixtureRequest,
         if cmdline_args is None:
             cmdline_args = []
         test_cmdline_args = self.global_extra_cmdline_args + cmdline_args
+        test_run_python_command = "--runpythontest"
+        if self.material_editor_test:
+            test_run_python_command = "--runpython"
```
[the remark this answers is not in the document]
Comment (jromnoa) [reply Tools/LyTestTools/ly_test_tools/o3de/editor_test.py:835]: There was a bug where it wasn't usable with `--runpythontest` and that should resolve the issue we are seeing with `.get_returncode()` (which is why the draft was opened)

File: Tools/LyTestTools/ly_test_tools/o3de/editor_test.py (hunk 5 above)
Comment (jromnoa) [reply Tools/LyTestTools/ly_test_tools/o3de/editor_test.py:351]: yep will be changing it to class inheritance from EditorTestSuite

File: AutomatedTesting/Gem/PythonTests/Atom/TestSuite_Main_GPU.py (hunk 4 above)
Comment (jromnoa) [reply AutomatedTesting/Gem/PythonTests/Atom/TestSuite_Main_GPU.py:161]: also in reply to your question about this being the final version, it is not, it was opened to help resolve a bug with `get_returncode()` which actually had to do with the `--runpythontest` command vs. the `--runpython` command which also wasn't working until this PR: https://github.com/o3de/o3de/pull/9136 

thank you for your feedback, will be updating it appropriately

Comment (jromnoa) [reply AutomatedTesting/Gem/PythonTests/Atom/TestSuite_Main_GPU.py:161]: I updated this but there isn't a specific base class for suites yet, but that will be in the next update (as well as declaring the GPU test correctly)

**Comment by jromnoa** [comment]
It should be non-GPU, I will update the `CMakeLists.txt` file.
Progress on this continues but still not complete, quite a hefty undertaking due to how the pytest collection can collide with some of the manual collection we are doing here. I just have a couple more bugs to work out then apply the changes I've made to `editor_test.py` to `material_editor_test.py` and this should hopefully finally be in a reviewable state and more closely match what the RFC laid out.

Disclaimer: I will not be using this RFC and task to do a complete re-factor of `editor_test.py` so expect some things to not be perfectly ideal and neatly clean. The main goal of this RFC is to support the MaterialEditor and I will stick to that goal, so some object inheritance or function signatures may seem strange or redundant. This is so the `test_suite_base.py` serves as a schematic for when the actual re-factor occurs in the future (a whole separate task with a whole separate effort involved).

File: AutomatedTesting/Gem/PythonTests/Atom/CMakeLists.txt [hunk 8]
```diff
@@ -100,4 +100,18 @@ if(PAL_TRAIT_BUILD_HOST_TOOLS AND PAL_TRAIT_BUILD_TESTS_SUPPORTED)
         COMPONENT
             Atom
     )
+    ly_add_pytest(
+        NAME AutomatedTesting::Atom_Main_MaterialEditor
+        TEST_SUITE main
+        TEST_REQUIRES gpu
+        TEST_SERIAL
+        TIMEOUT 700
+        PATH ${CMAKE_CURRENT_LIST_DIR}/TestSuite_Main_Null_Render_MaterialEditor.py
```
[the remark this answers is not in the document]
Comment (jromnoa) [reply AutomatedTesting/Gem/PythonTests/Atom/CMakeLists.txt:109]: It should be non-GPU, I will update the CMakeLists.txt file.

[the remark this answers is not in the document]
Comment (jromnoa) [reply AutomatedTesting/Gem/PythonTests/Atom/CMakeLists.txt:109]: As a side note, updating the existing tests will be a separate task. The goal of this task is to support MaterialEditor with the new `editor_test.py` framework tools.

File: AutomatedTesting/Gem/PythonTests/Atom/tests/MaterialEditor_Atom_LaunchMaterialEditor.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    import Atom.atom_utils.material_editor_utils as material_editor
+
+    from editor_python_test_tools.utils import Report, Tracer, TestHelper
+
+    with Tracer() as error_tracer:
+        # 1. Verify Material Inspector pane visibility to confirm the MaterialEditor launched.
+        material_editor.set_pane_visibility("Inspector", 1)
+        Report.result(
+            Tests.material_editor_launched,
+            material_editor.is_pane_visible("Inspector") == 1)
```
[the remark this answers is not in the document]
Comment (jromnoa) [reply AutomatedTesting/Gem/PythonTests/Atom/tests/MaterialEditor_Atom_LaunchMaterialEditor.py:40]: will add this as part of the new PR im putting together right now

File: AutomatedTesting/Gem/PythonTests/Atom/tests/MaterialEditor_Atom_LaunchMaterialEditor.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    :return: None
+    """
+
+    import Atom.atom_utils.material_editor_utils as material_editor
+
+    from editor_python_test_tools.utils import Report, Tracer, TestHelper
+
+    with Tracer() as error_tracer:
+        # 1. Verify Material Inspector pane visibility to confirm the MaterialEditor launched.
+        material_editor.set_pane_visibility("Inspector", 1)
```
[the remark this answers is not in the document]
Comment (jromnoa) [reply AutomatedTesting/Gem/PythonTests/Atom/tests/MaterialEditor_Atom_LaunchMaterialEditor.py:37]: same as below, will add this to new PR

**Comment by jromnoa** [comment]
New PR can be found here: https://github.com/o3de/o3de/pull/9609

**Pull request closed without merging**